Preparation for FontCustomPlatformData should be deserialized with SafeFontParser when in LDM (271976)

vitorroriz opened the pull request:
#### aec6b7b9922a9106ca4a4b9a946b39e321968354
<pre>
Preparation for FontCustomPlatformData should be deserialized with SafeFontParser when in LDM (271976)
<a href="https://bugs.webkit.org/show_bug.cgi?id=271976">https://bugs.webkit.org/show_bug.cgi?id=271976</a>
<a href="https://rdar.apple.com/125729105">rdar://125729105</a>

Reviewed by Alex Christensen.

FontCustomPlatformData deserialization requires font parsing and we should
use SafeFontParser if we are in lockdown mode (LDM).

We need some context to know if we are in LDM. For that reason,
the &quot;CacheFontCustomPlatformData&quot; message received by the GPUP
will no longer received a deserialized FontCustomPlatformData.

Instead, it will receive a FontCustomPlatformSerializedData and
the message handler, which knows if we are in LDM, will deserialize
FontCustomPlatformSerializedData.

This patch should not change behavior, it is only a preparation
for making the data path adequate to the future change.

* Source/WebCore/platform/graphics/FontCustomPlatformData.h:
* Source/WebCore/platform/graphics/coretext/FontCustomPlatformDataCoreText.cpp:
(WebCore::FontCustomPlatformData::tryMakeFromSerializationData):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
(WebKit::GPUConnectionToWebProcess::isLockdownSafeFontParserEnabled const):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::cacheFontCustomPlatformData):
(WebKit::RemoteRenderingBackend::shoulUsedLockdownFontParser const):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/GPUProcessPreferencesForWebProcess.h:
* Source/WebKit/Shared/GPUProcessPreferencesForWebProcess.serialization.in:
* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::preferencesForGPUProcess const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::cacheFontCustomPlatformData):

Canonical link: <a href="https://commits.webkit.org/277060@main">https://commits.webkit.org/277060@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d3879e9b0219a26c55a3a921c2dd2c5b88be7cd0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46478 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25634 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49082 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49154 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42519 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48785 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29994 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23097 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37924 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47056 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22659 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40087 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19174 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/46344 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/20038 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41191 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4523 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/42807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41567 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50992 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21483 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17925 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45177 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/22775 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44116 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10292 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/23183 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22478 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ddf56a7f181edfa4a9495e293079542e2552964c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/168c01c5d955b0fb595d523e36c08dee35971bc9. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f985ba5111badac51393121ee66c17ceb51cb33c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9226c45be83eab37c30b52b33d637f020c6a4e44. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26687

achristensen07 commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
```diff
@@ -364,10 +364,16 @@ void RemoteRenderingBackend::cacheFont(const Font::Attributes& fontAttributes, F
-    m_remoteResourceCache.cacheFontCustomPlatformData(WTFMove(customPlatformData));
+
+    // FIXME: rdar://124235570 Get LDM state from GPU connection when feature is ready.
+    constexpr bool  isInLockdownMode { false };
```
nit: extra space.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2094c4ef646889c6b67d820173eede2ffeba9005. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a558e150078b8d01829d5def13dffc67fe8d1adc. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/78068e26b9fed2bd358a870177b65dc180843f81. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/46e2a02c3ca9285151a12a118dbe8650f55d2898. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a1c84d5d03c78f4e6e11b5267e1334d8ba15546d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c3b15b58745fef5cada8bc426166323b30fe99b7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ff7eca7c786fec78f1991a05009363fb4aaee116. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0f42bdfc70c702c737388048deb366b2751d6337. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d12cdf48390596c9235cb0dccbca5341ed5f0eb3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26687

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d3879e9b0219a26c55a3a921c2dd2c5b88be7cd0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26687

achristensen07 commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
```diff
@@ -160,6 +160,8 @@ class GPUConnectionToWebProcess
     PAL::SessionID sessionID() const { return m_sessionID; }
 
     bool isLockdownModeEnabled() const { return m_isLockdownModeEnabled; }
+    bool isLockdownSafeFontParserEnabled() const { return m_preferences.isLockdownSafeFontParserEnabled; }
```
Do these need to be separate values?  Can they ever be different?

vitorroriz replied:
Yes, they are independent. The SafeFontParser can be disabled even in during lockdown mode, since it's controlled by an experimental feature flag.

vitorroriz committed:
Preparation for FontCustomPlatformData should be deserialized with SafeFontParser when in LDM (271976)
https://bugs.webkit.org/show_bug.cgi?id=271976
rdar://125729105

Reviewed by Alex Christensen.

FontCustomPlatformData deserialization requires font parsing and we should
use SafeFontParser if we are in lockdown mode (LDM).

We need some context to know if we are in LDM. For that reason,
the "CacheFontCustomPlatformData" message received by the GPUP
will no longer received a deserialized FontCustomPlatformData.

Instead, it will receive a FontCustomPlatformSerializedData and
the message handler, which knows if we are in LDM, will deserialize
FontCustomPlatformSerializedData.

This patch should not change behavior, it is only a preparation
for making the data path adequate to the future change.

* Source/WebCore/platform/graphics/FontCustomPlatformData.h:
* Source/WebCore/platform/graphics/coretext/FontCustomPlatformDataCoreText.cpp:
(WebCore::FontCustomPlatformData::tryMakeFromSerializationData):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
(WebKit::GPUConnectionToWebProcess::isLockdownSafeFontParserEnabled const):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::cacheFontCustomPlatformData):
(WebKit::RemoteRenderingBackend::shoulUsedLockdownFontParser const):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/GPUProcessPreferencesForWebProcess.h:
* Source/WebKit/Shared/GPUProcessPreferencesForWebProcess.serialization.in:
* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::preferencesForGPUProcess const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::cacheFontCustomPlatformData):

Canonical link: https://commits.webkit.org/277060@main

webkit-commit-queue commented:
Committed 277060@main (aec6b7b9922a): <https://commits.webkit.org/277060@main>

Reviewed commits have been landed. Closing PR #26687 and removing active labels.
